packager/file/udp_file.cc

```diff
@@ -43,6 +47,85 @@ bool IsIpv4MulticastAddress(const struct in_addr& addr) {
   return (ntohl(addr.s_addr) & 0xf0000000) == 0xe0000000;
 }
 
+bool IsIpv6MulticastAddress(const struct in6_addr& addr) {
+  if (addr.s6_addr[0] == 0xff) {
+    return true;
+  }
+
+  return false;
+}
+
+template <typename T>
+bool SpecificMalticast(int ai_family,
+                       int level,
+                       void *imr_interface,
+                       void *imr_sourceaddr,
+                       SOCKET socket,
+                       T& source_multicast_group,
+                       std::unique_ptr<UdpOptions>&& options) {
+  if (inet_pton(ai_family,
+                options->interface_address().c_str(),
+                imr_interface) != 1) {
+    LOG(ERROR) << "Malformed IP interface address "
+               << options->interface_address();
+    return false;
+  }
+  if (inet_pton(ai_family,
+                options->source_address().c_str(),
+                imr_sourceaddr) != 1) {
+    LOG(ERROR) << "Malformed IP source specific multicast address "
+               << options->source_address();
+    return false;
+  }
+  if (setsockopt(socket,
+                 level,
+                 IP_ADD_SOURCE_MEMBERSHIP,
+                 reinterpret_cast<const char*>(&source_multicast_group),
+                 sizeof(source_multicast_group)) < 0) {
+    LOG(ERROR) << "Failed to join multicast group.";
+    return false;
+  }
+
+  return true;
+}
+
+template <typename T>
+bool Multicast(int ai_family,
+               int level,
+               void *imr_interface,
+               SOCKET socket,
+               T& multicast_group,
+               std::unique_ptr<UdpOptions>&& options) {
+    if (inet_pton(ai_family,
+                  options->interface_address().c_str(),
+                  imr_interface) != 1) {
+      LOG(ERROR) << "Malformed IP interface address "
+                 << options->interface_address();
+      return false;
+    }
+    if (setsockopt(socket, level, IP_ADD_MEMBERSHIP,
```

Comment:
Similar to above, does it with IPv6?

According to https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.1.0/com.ibm.zos.v2r1.hale001/ipv6d0141001708.htm, the IPv6 equivalent is IPV6_JOIN_GROUP. There is also a protocol independent version: MCAST_JOIN_GROUP.